src/main/java/chapter01/factory/DaoFactory.java

```diff
@@ -0,0 +1,45 @@
+package chapter01.factory;
+
+import chapter01.connection.ConnectionMaker;
+import chapter01.connection.CountingConnectionMaker;
+import chapter01.connection.DConnectionMaker;
+import chapter01.connection.SimpleConnectionMaker;
+import chapter01.dao.AccountDao;
+import chapter01.dao.MessageDao;
+import chapter01.dao.UserDao;
+import com.mysql.jdbc.Driver;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.jdbc.datasource.SimpleDriverDataSource;
+
+import javax.sql.DataSource;
+
+@Configuration
+public class DaoFactory {
+
+    @Bean
+    public UserDao userDao() {
+        return new UserDao(dataSource());
```

Comment:
요런 방식의 메소드 호출을 통한 종속성 주입보다는, 정말 의존관계의 결정이 IoC Container에 위임될 수 있도록 아래와 같은 방법을 선호하는 편입니다.

```suggestion
    public UserDao userDao(DataSource dataSource) {
        return new UserDao(dataSource);
    }
```

요렇게 하면 어떤 이점이 있을지 생각해 보시죠.

Reply:
DataSource를 외부에서 주입 받을 경우 
1. 확장성 가능해진다.
외부에서 데이터소스 구성 정보를 변경하여도 코드 변경 없이 구성 정보를 수정하거나 확장할 수 있습니다.
2. 테스트 가능해진다.
userDao 유닛 테스트가 가능해집니다.
3. 운영 관리를 외부에서 가능하다
Product환경, Release 환경 등 여러 환경에 대한 설정을 외부에서 관리가 가능합니다.